wee_alloc/src/imp_static_array.rs

```diff
@@ -1,21 +1,40 @@
-use const_init::ConstInit;
 use super::AllocErr;
+use const_init::ConstInit;
 #[cfg(feature = "extra_assertions")]
 use core::cell::Cell;
 use core::ptr::NonNull;
 use memory_units::{Bytes, Pages};
 use spin::Mutex;
 
-const SCRATCH_LEN_BYTES: usize = include!(concat!(env!("OUT_DIR"), "/wee_alloc_static_array_backend_size_bytes.txt"));
+const SCRATCH_LEN_BYTES: usize = include!(concat!(
+    env!("OUT_DIR"),
+    "/wee_alloc_static_array_backend_size_bytes.txt"
+));
 static mut SCRATCH_HEAP: [u8; SCRATCH_LEN_BYTES] = [0; SCRATCH_LEN_BYTES];
```

Comment:
Another thing we should do to avoid most scenarios where we don't have proper alignment is to force an alignment on the `SCRATCH_HEAP` via

```rust
#[repr(align(4096)]
struct ScratchHeap([u8; SCRATCH_LEN_BYTES] = [0; SCRATCH_LEN_BYTES]);

static mut SCRATCH_HEAP: ScratchHeap = ScratchHeap([0; SCRATCH_LEN_BYTES]);
```

Reply:
In fact, is just this change enough to completely fix the problem, since the only call to `alloc_pages` ensures that the number of pages is guaranteed to be big enough to contain a cell of the requested alignment? Which extra assertion in particular is failing right now? Is it `assert_is_word_aligned`? If so, then I think just this change is good enough.

Reply:
I have added this as well to the upcoming push, and it does seem to be the ultimate fix but I have one possibly unfounded reservation. Once I implemented the above rough alignment fix no extra assertions failed. Prior to the fix the extra_assertion feature forced failure at around the first allocation attempt,  I believe somewhere in `alloc_first_fit` in the call to 
`assert_aligned_to(allocated.data(), align);` Looking at `assert_is_word_aligned` it is a wrapper for `assert_aligned_to` so yes potentially just this change is good enough. 

However, I had instrumented your code to speed up problem source discovery, and I recall noticing that somewhere in the allocation process, the resulting size request was not always a multiple of the byte alignment requirement. I (also foolishly) reverted all of that instrumentation to submit a nice clean PR and do not exactly recall where, so potentially this is not an issue. But for instance it may have been when, in `new_cell_for_free_list`, the size of the requested allocation is compared to and changed based on whether the value ensures the minimum cell size is left over after allocation (e.g. in `new_cell_for_free_list` in lib.rs:780  `let size: Bytes = cmp::max(size.into(), (align + Self::MIN_CELL_SIZE) * Words(2));` ). 

Im not sure without instrumenting the code again, but is it possible that the end resulting size request may result in `*offset` becoming unaligned so that the next allocation will be unaligned even if the `SCRATCH_HEAP` is forced to align as a page boundary? If this is possible then Im not sure just this fix would solve the issue. I can look into this at some point if you would like?

Reply:
This code *should* be ensuring that the requested number of pages allocated should always be able to satisfy the allocation request, if `alloc_pages` succeeds: https://github.com/rustwasm/wee_alloc/blob/master/wee_alloc/src/lib.rs#L776-L783

The allocation doesn't have to be a multiple of the alignment, just has to be big enough that an aligned chunk of the requested size must exist within it somewhere.